src/backend/lang_c/c_simple_ast.rs

```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
+        let mut is_valid = true;
+        let mut ret = String::new();
+        let nodes = cast.ast.node_indices();
+        for node in nodes {
+            if let Err(msg) = verifier(node, cast) {
+                ret = format!("{}; {}", ret, msg);
+                is_valid = false;
+            }
+        }
+        if is_valid {
+            Ok(())
+        } else {
+            Err(format!("{} @ {}", ret, name.to_string()))
+        }
+    }
+
+    // 1. There are at most 1 ActionEdge::Normal from each node.
+    fn verify_normal_action(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(_)) => {},
+            _ => return Ok(()),
+        };
+        let normal_actions = cast.ast.edges_directed(node, Direction::Outgoing)
+            .into_iter()
+            .filter(|e| {
+                match e.weight() {
+                    &SimpleCASTEdge::Action(ActionEdge::Normal) => true,
+                    _ => false,
+                }
+            }).collect::<Vec<_>>();
+        if normal_actions.len() <= 1 {
+            Ok(())
+        } else {
+            Err(format!("There are {} next normal actions @ {:?}",
+                        normal_actions.len(), node))
+        }
+    }
+
+    // 2. There are ActionEdge::IfThen, ValueEdge::Conditional from ActionNode::If
+    fn verify_if(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(ActionNode::If)) => {},
+            _ => return Ok(()),
+        };
+        let mut ret = String::new();
+        let cond = cast.branch_condition(node);
+        if cond.is_none() {
+            ret = format!("{}; No condition node is found.", ret);
+        }
+        let branches = cast.branch(node).map(|x| x.clone());
+        if branches.is_none() {
+            ret = format!("{}; No branch is found.", ret);
+        }
+        if cond.is_none() || branches.is_none() {
+            Err(ret)
+        } else {
+            Ok(())
+        }
+    }
+
+    // 3. The targets of value edges are ValueNode, The target of action edges are ActionNode.
+    fn verify_edge_action(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        let mut is_err = false;
+        let mut ret = String::new();
+        let edges = cast.ast
+            .edges_directed(node, Direction::Outgoing).collect::<Vec<_>>();
+        for edge in edges {
+            match (edge.weight(), cast.ast.node_weight(edge.target())) {
+                (SimpleCASTEdge::Action(_), Some(&SimpleCASTNode::Action(_))) => {},
+                (SimpleCASTEdge::Value(_), Some(&SimpleCASTNode::Value(_))) => {},
+                (SimpleCASTEdge::Action(ActionEdge::GotoDst), Some(&SimpleCASTNode::Entry)) => {},
+                _ => {
+                    is_err = true;
+                    ret = format!("{}; {:?} {:?}", ret, edge.weight(),
```

Comment:
Again, `ret` may be empty before this, in which case `ret` will start with `"; ..."`.